### PR TITLE
Touch on ast/simplify

### PR DESCRIPTION
**Pull request by povik**
This is supposed to untangle the simplify code a bit by removing three of its arguments (in_lvalue/in_param/at_zero), and making few other changes.

I expect the at_zero removal to be less objectionable than the in_param/in_lvalue removal, since the latter (with the way it's being replaced) requires the code to be sprinkled with calls to `fixup_hierarchy_flags` and `set_attribute` (new methods). That may not be as bad if we (1) don't require the in_param/in_lvalue values of the old code to be perfectly reproduced (which the change here does reproduce, at least as exercised by `make test`); (2) move to a coding style in the simplify code that doesn't rely on explicitly modifying the children list of an existing node as much. See the commits and let me know what you think.

This should be passing `make test` at each commit.

### CODE REVIEW

File: frontends/ast/simplify.cc
Comment (povik) [inline frontends/ast/simplify.cc]: I want to point out these asserts (comparing the flags on the node with the flags passed in to simplify) are upheld through `make test`.

**Comment by zachjs** [comment]
Thank you for your submission! I think this offers some nice cleanup overall, though I'd like to hear from @jix and/or @clairexen before moving forward. I think we're far more likely to get this merged if it each commit is split into a separate PR. So far, I have only skimmed a bit and provided some initial thoughts.

* The first commit seems straightforwardly good and mergable as-is.
* This seems like a nice improvement, but it is a sweeping stylistic change that A) requires broader agreement; and B) is essentially orthogonal to the rest of the PR.
* This seems like a nice strategy. You mention assuming `id2ast` is valid, but I'm not sure how we benefit from making this assumption. Could `clone_at_zero` look up identifiers in `current_scope` as was done before? Given the subtree has no scoping, I don't see any harm in using the scope as a fallback.
* I like splitting up obviating the parameter and actually removing it. However, I would really like to see the _prior_ commit remove the code that _uses_ `at_zero`, rather than making it unreachable there and removing it here. This would make it easier to compare the new "at zero" logic with what it replaces, both now and in the future.
* I haven't had the chance to look closely at the remaining three commits, but they may be trickier and/or more contentious.

**Comment by clairexen** [comment]
I agree with everything zach said, and I'm fine with merging the first few simple PRs first, and then discussing the remaining commits as individual PRs.

**Comment by povik** [comment]
Hi Zach, Claire,

dropped the last three commits, so now there are only those Zach commented on. Also did rebase.

>  You mention assuming `id2ast` is valid, but I'm not sure how we benefit from making this assumption. Could `clone_at_zero` look up identifiers in `current_scope` as was done before? Given the subtree has no scoping, I don't see any harm in using the scope as a fallback.

I think it just seemed favorable to rely on an attribute of the AST node in the node's method, rather than be coupled to the simplify() process through the `current_scope` variable. Bit surprisingly, when I now tried to use `current_scope` directly to look up identifiers in `clone_at_zero`, I started failing some tests, even though that's akin to what the original at_zero simplify code does. I didn't investigate, and changed it to use `id2ast` and fall back on `current_scope` for now -- that's what's in the PR now.

> I like splitting up obviating the parameter and actually removing it. However, I would really like to see the _prior_ commit remove the code that _uses_ `at_zero`, rather than making it unreachable there and removing it here. This would make it easier to compare the new "at zero" logic with what it replaces, both now and in the future.

Changed it to be that way. Prior commit replaces the code conditioned on `at_zero` with an assertion that the flag is false, latter commit removes the flag.

File: frontends/ast/simplify.cc
Comment (zachjs) [inline frontends/ast/simplify.cc:837]: AFAIK, this backslash isn't required, and isn't used elsewhere in this file.

File: frontends/ast/simplify.cc
Comment (zachjs) [inline frontends/ast/simplify.cc:49]: I'm partial to defining this ins `ast.cc` instead.

File: frontends/ast/simplify.cc
Comment (zachjs) [inline frontends/ast/simplify.cc:799]: Let's drop the backslashes.

**Review by zachjs — APPROVED** [review]
I'm fine with merging these 4 commits as a batch. I plan to merge this shortly _after_ the next release is cut. I have a few very minor nits from my second review. I'm happy to fix them if you 1) don't object; and 2) don't have the chance address them in the next few days.

Thanks again for your contribution! These are great simplifications.

**Comment by povik** [comment]
Hi Zach,

thanks for taking another look at the changes! I happen to be AFK for the next week but feel free to merge it with those minor adjustments.

**Pull request merged**
